cpp/include/cudf/detail/utilities/strong_index.hpp

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/types.hpp>
+
+namespace cudf {
+
+namespace detail {
+
+enum class index_type_name {
+  LEFT,
+  RIGHT,
+};
+
+template <index_type_name IndexName>
+struct strong_index {
```

Comment:
Is all this needed? 

I had assumed that these would just need to be very minor strong type wrappers to select the right overload.

e.g., in the issue I'd suggested:
```
enum class lhs_index_t : int32_t{};
enum class rhs_index_t : int32_t{};
```

Reply:
I thought that was not meant to be a literal suggestion, but an oversimplified case for the purpose of the demo. If it turns out that operators aren't needed, I'd be happy to use that instead.

Reply:
> I thought that was not meant to be a literal suggestion, but an oversimplified case for the purpose of the demo/concept. 

That's a fair assumption :) 

Though in this case, it was meant to be more or less literal. The strong type just selects the overload and then we can cast it to an `int`.

Reply:
I thought the same---why not just `enum class lhs_index_t : int32_t{};`? In the row comparator, just cast the input index into `int`. No operations should be performed directly on the `_index_t`.